File: OC/FlowLayout/MBFlowLayout.m
//
//  MBFlowLayout.m
//  FlowLayout
//
//  Created by MaBiao on 2018/3/29.
//  Copyright © 2018年 MaBiao. All rights reserved.
//

#import "MBFlowLayout.h"
//列间距
static const CGFloat defaultColMargin = 10;
//行间距
static const CGFloat defaultRowMargin = 10;
//默认列数
static const CGFloat defaultColCount = 3;
//边缘间距
static const UIEdgeInsets defaultEdgeInsets = {10,10,10,10};


@interface MBFlowLayout()

/** 存放所有cell的布局属性 */
@property (nonatomic, strong) NSMutableArray *attrsArray;
/** 存放每列最大的Y值 */
@property (nonatomic, strong) NSMutableArray *columMaxHeights;

- (CGFloat)rowMargin;
- (CGFloat)columnCount;
- (CGFloat)columnMargin;
- (UIEdgeInsets)edgeInsets;
@end

@implementation MBFlowLayout

#pragma mark - 数据处理
- (CGFloat)rowMargin
{
    if ([self.delegate respondsToSelector:@selector(numOfRowMarginInFlowLayout:)]) {
        return [self.delegate numOfRowMarginInFlowLayout:self];
    }else{
        return defaultRowMargin;
    }
}
- (CGFloat)columnCount
{
    if ([self.delegate respondsToSelector:@selector(columnCountInFlowLayout:)]) {
        return [self.delegate columnCountInFlowLayout:self];
    }else{
        return defaultColCount;
    }
}
- (CGFloat)columnMargin
{
    if ([self.delegate respondsToSelector:@selector(columnMarginInFlowLayout:)]) {
        return [self.delegate columnMarginInFlowLayout:self];
    }else{
        return defaultColMargin;
    }
}
- (UIEdgeInsets)edgeInsets
{
    if ([self.delegate respondsToSelector:@selector(edgeInsetsInFlowLayout:)]) {
        return [self.delegate edgeInsetsInFlowLayout:self];
    }else{
        return defaultEdgeInsets;
    }
}

- (NSMutableArray *)columMaxHeights
{
    if (!_columMaxHeights) {
        _columMaxHeights = [NSMutableArray array];
    }
    return _columMaxHeights;
}
- (NSMutableArray *)attrsArray
{
    if (!_attrsArray) {
        _attrsArray = [NSMutableArray array];
    }
    return _attrsArray;
}

//初始化，每次布局都调用
- (void)prepareLayout
{
    [super prepareLayout];
    
    //重置每一列的最大Y值
    [self.columMaxHeights removeAllObjects];
    for (NSInteger i = 0; i < self.columnCount; i++) {
        [self.columMaxHeights addObject:@(self.edgeInsets.top)];
    }
    
    //清除之前所有的布局属性
    [self.attrsArray removeAllObjects];
 
    NSInteger count = [self.collectionView numberOfItemsInSection:0];
    
    for (NSInteger i = 0; i < count; i++) {
        NSIndexPath *indexPath = [NSIndexPath indexPathForItem:i inSection:0];
        UICollectionViewLayoutAttributes *attr = [self layoutAttributesForItemAtIndexPath:indexPath];
        [self.attrsArray addObject:attr];
    }
}
//决定cell排布
- (NSArray<UICollectionViewLayoutAttributes *> *)layoutAttributesForElementsInRect:(CGRect)rect
{
    return self.attrsArray;
}
//返回indexPath位置的cell对应的布局属性
- (UICollectionViewLayoutAttributes *)layoutAttributesForItemAtIndexPath:(NSIndexPath *)indexPath
{
    //创建布局属性
    UICollectionViewLayoutAttributes *attr = [UICollectionViewLayoutAttributes layoutAttributesForCellWithIndexPath:indexPath];

    CGFloat collectionW = self.collectionView.frame.size.width;
    
    CGFloat w = (collectionW - self.edgeInsets.left - self.edgeInsets.right - (self.columnCount - 1)* self.columnMargin) /  self.columnCount;
    
    //利用代理，外部设置高度
    CGFloat h = [self.delegate flowLayout:self heightForItemAtIndex:indexPath.item itemWidth:w];
    
    //找出最大Y值最小的那一列
    CGFloat minHeight = [self.columMaxHeights[0] doubleValue];//默认第一列最短
    NSInteger destColum = 0;//最短的那一列
    for (int i = 1; i < self.columnCount; i++) {
        CGFloat columMaxHeight = [self.columMaxHeights[i] doubleValue];
        if (minHeight > columMaxHeight) {
            minHeight = columMaxHeight;
            destColum = i;
        }
    }
    
    CGFloat x = self.edgeInsets.left + destColum * (w + self.columnMargin);
    CGFloat y = minHeight;
    //第一列处理
    if (y != self.edgeInsets.top) {
        y += self.rowMargin;
    }
    
    attr.frame  = CGRectMake(x, y, w, h);
    
    //更新最短那列的高度
    self.columMaxHeights[destColum] = @(CGRectGetMaxY(attr.frame));
    
    return attr;
}

- (CGSize)collectionViewContentSize
{
    // 找出最长那一列的最大Y值
    CGFloat destMaxY = [self.columMaxHeights[0] doubleValue];
    for (int i = 1; i < self.columnCount; i++) {
        // 取出第i列的最大Y值
        CGFloat columMaxY = [self.columMaxHeights[i] doubleValue];
        if (destMaxY < columMaxY) {
            destMaxY = columMaxY;
        }
    }
    return CGSizeMake(0, destMaxY + self.edgeInsets.bottom);
}

@end

File: Swift/FlowLayout/FlowLayout/MBFlowLayout.swift
//
//  MBFlowLayout.swift
//  FlowLayout
//
//  Created by Telit on 2018/7/13.
//  Copyright © 2018年 iMaBiao. All rights reserved.
//

import UIKit

@objc protocol MBFlowLayoutDelegate : class {
    
    //外部设置布局的高度（必选）
    func flowLayoutHeightForItemAtIndex(_ flowLayout: MBFlowLayout, _ index: Int , _ itemWidth: CGFloat) -> CGFloat
    
    //可选
    @objc optional func columnCountInFlowLayout(_ flowLayout: MBFlowLayout) -> CGFloat
    @objc optional func numOfRowMarginInFlowLayout(_ flowLayout: MBFlowLayout) -> CGFloat
    @objc optional func columnMarginInFlowLayout(_ flowLayout: MBFlowLayout) -> CGFloat
    @objc optional func edgeInsetInFlowLayout(_ flowLayout: MBFlowLayout) -> UIEdgeInsets
}


class MBFlowLayout: UICollectionViewLayout {

    weak var delegate : MBFlowLayoutDelegate?
    
    ///存放所有cell的布局属性
    lazy var attrsArray : [UICollectionViewLayoutAttributes] =  Array()
    ///存放每列最大的Y值
    lazy var columMaxHeights: [CGFloat] = Array()
    
    var columnCount: CGFloat {
        guard let count = delegate?.columnCountInFlowLayout!(self) else {
            return 3
        }
        return count
    }
    
    var edgeInsets: UIEdgeInsets {
        
        guard let edge = delegate?.edgeInsetInFlowLayout!(self)  else {
            return UIEdgeInsetsMake(10, 10, 10, 10)
        }
        return edge
    }
    
    var columnMargin: CGFloat {
        guard let margin = delegate?.columnMarginInFlowLayout!(self) else {
            return 10
        }
        return margin
    }
    var rowMargin: CGFloat {
        
        guard let margin = delegate?.numOfRowMarginInFlowLayout!(self) else {
            return 10
        }
        return margin
    }
    
    
    override func prepare() {

        super.prepare()

        columMaxHeights.removeAll()

        for _ in 0..<Int(columnCount) {
            columMaxHeights.append(edgeInsets.top)
        }

        attrsArray.removeAll()

        guard let count = collectionView?.numberOfItems(inSection: 0) else {
            return
        }

        for i  in 0..<count {
            let indexPath = NSIndexPath(item: i, section: 0)as IndexPath?
            guard let attr = self.layoutAttributesForItem(at: indexPath! ) else{
                return
            }
            attrsArray.append(attr)
        }
    }
    
    //决定cell排布
    override func layoutAttributesForElements(in rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
            return attrsArray
    }
    
    //返回indexPath位置的cell对应的布局属性
    override func layoutAttributesForItem(at indexPath: IndexPath) -> UICollectionViewLayoutAttributes? {
      
        //创建布局属性
        let attr = UICollectionViewLayoutAttributes(forCellWith: indexPath)
        
        guard let collectionW = collectionView?.frame.size.width else {
            return nil
        }

        let temp = (columnCount - 1) * columnMargin
        
        let w = (collectionW - edgeInsets.left - edgeInsets.right - temp ) / columnCount
        
        guard let h: CGFloat = delegate?.flowLayoutHeightForItemAtIndex(self, indexPath.item, w) else {
            return nil
        }
        
        var minHeight: CGFloat = columMaxHeights[0]//默认第一列最短
        var destColum = 0   //最短的那一列
        
        for i in 0..<Int(columnCount) {
            let columMaxHeight: CGFloat = columMaxHeights[i]
            if minHeight > columMaxHeight {
                minHeight = columMaxHeight
                destColum = i ;
            }
        }
        
        let x = edgeInsets.left + CGFloat(destColum) * (w + columnMargin)
        var y = minHeight
        //第一列处理
        if y != edgeInsets.top {
            y += rowMargin
        }
        attr.frame = CGRect(x: x, y: y, width: w, height: h)
         //更新最短那列的高度
        columMaxHeights[destColum] = attr.frame.maxY
        return attr
    }
    
    override var collectionViewContentSize: CGSize{
        
        var destMaxY : CGFloat = columMaxHeights[0]
        for i in 0..<Int(columnCount) {
            let columMaxY = columMaxHeights[i]
            if destMaxY < columMaxY {
                destMaxY = columMaxY
            }
        }
        return CGSize(width: 0, height: destMaxY + edgeInsets.bottom)
    }
    
}
